### ISSUE TITLE
Do faceted fields need to be indexed? Do non-indexable fields need to be marked as optional?

### ISSUE THREAD

**Issue by kdevan** [typesense/typesense#416]
## Questions

1. Do faceted fields need to be indexed? I have a case where I have an attribute display name and attribute slug. I'd like to index the attribute display name so that's it's searchable with the document. I don't want the attribute slug to be searchable but I do want to use it as a facet. Is there a way to accomplish this?
2. When I mark a field as non-indexable, I get an error that I also need to mark it as optional. In the docs I see fields marked as non-indexable without being marked as optional. Is this the client library I'm using or is that correct, that all non-indexable fields must be marked as optional?

Just want to take this chance to say thank you to everyone who contributes to this project. It's been an amazing experience setting this up. I need to index about 10 million documents and Typesense handles this beautifully while an alternative I tried struggled. Amazing work!

**Comment by kishorenc** [comment]
@kdevan 

Thank you, happy to hear that.

1. Yes, facet fields have to be indexed. There is currently no way to make a field be indexable only for facet but not search. Happy to consider this as a feature request.
2. This is probably a bug, likely an outcome of the `index` property being designed along with automatic schema, which requires fields to be optional. I will look into this. I presume you're using Typesense v0.21?

**Comment by unclescooby** [comment]
I too noticed a field marked as non-indexable must be set to optional. Regex/wildcard fields must be optional, and they spawn new fields that are also optional. However, in reverse, a field that is optional does not have to be non-indexable per se!!! In fact, the last time I checked (v0.21) the wildcard field had to be indexable (the regex field did not).

**Comment by miguel-petrarca** [comment]
any timeline for resolution of this?

**Comment by victorkristof** [comment]
I found out that is you set the type of a field as `auto` then you can make it un-indexable. It indeed also works if you make the field optional. In my case it was safer to set the type to auto rather than making the field optional.

It would be great to be able to define both the type and `'index': False` without making it optional.

**Comment by vegarsti** [comment]
> I too noticed a field marked as non-indexable must be set to optional. Regex/wildcard fields must be optional, and they spawn new fields that are also optional. However, in reverse, a field that is optional does not have to be non-indexable per se!!! In fact, the last time I checked (v0.21) the wildcard field had to be indexable (regex fields did not).
> 
> EDIT: re facets I can actually create a field like `["name" => "foo", "type" => "string", "optional" => true, "index" => false, "facet" => true],` without error, so I'm wondering if that's a bug, or if typesense simply overrides the index parameter (I cant check if typesense indexes that field to memory or just stores to disk).

At least as of vX, it's no longer possible to have a field that is optional and a facet, but not indexed:

```
> curl "http://localhost:8108/collections" \
       -X POST \
       -H "Content-Type: application/json" \
       -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
       -d '{
         "name": "companies2",
         "fields": [
           {"name": "company_name", "type": "string", "index": true },
           {"name": "num_employees", "type": "int32", "index": true },
           {"name": "country", "type": "string", "facet": true, "index": false, "optional": true }
         ],
         "default_sorting_field": "num_employees"
       }'
{"message": "Field `country` cannot be a facet since it's marked as non-indexable."}
```

**Comment by kishorenc** [comment]
@vegarsti 

The property `index: false` disables all types of in-memory indexing and this includes facet also. So this is expected behavior.

The other issue of a field not being allowed to be both `index: false` AND `optional: false` is fixed in 0.26 RC builds.

**Comment by vegarsti** [comment]
> @vegarsti
> 
> The property `index: false` disables all types of in-memory indexing and this includes facet also. So this is expected behavior.
> 
> The other issue of a field not being allowed to be both `index: false` AND `optional: false` is fixed in 0.26 RC builds.

Yes, I meant to show that this was resolved.
